packages/device_info_plus/tizen/src/device_info_plus_tizen_plugin.cc

```diff
@@ -0,0 +1,295 @@
+// Copyright 2021 Samsung Electronics Co., Ltd. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "device_info_plus_tizen_plugin.h"
+
+#include <flutter/event_channel.h>
+#include <flutter/event_sink.h>
+#include <flutter/event_stream_handler_functions.h>
+#include <flutter/method_channel.h>
+#include <flutter/plugin_registrar.h>
+#include <flutter/standard_method_codec.h>
+#include <system_info.h>
+
+#include <map>
+#include <memory>
+#include <sstream>
+#include <string>
+
+#include "log.h"
+
+class DeviceInfoPlusTizenPlugin : public flutter::Plugin {
+ public:
+  static void RegisterWithRegistrar(flutter::PluginRegistrar *registrar) {
+    auto channel =
+        std::make_unique<flutter::MethodChannel<flutter::EncodableValue>>(
+            registrar->messenger(), "dev.fluttercommunity.plus/device_info",
+            &flutter::StandardMethodCodec::GetInstance());
+
+    auto plugin = std::make_unique<DeviceInfoPlusTizenPlugin>();
+
+    channel->SetMethodCallHandler(
+        [plugin_pointer = plugin.get()](const auto &call, auto result) {
+          plugin_pointer->HandleMethodCall(call, std::move(result));
+        });
+
+    registrar->AddPlugin(std::move(plugin));
+  }
+
+  DeviceInfoPlusTizenPlugin() {}
+
+  virtual ~DeviceInfoPlusTizenPlugin() {}
+
+ private:
+  void HandleMethodCall(
+      const flutter::MethodCall<flutter::EncodableValue> &method_call,
+      std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>> result) {
+    LOG_INFO("method : %s", method_call.method_name().data());
+
+    char *value = nullptr;
+    flutter::EncodableMap msg;
+    int ret;
+    bool hasResult = false;
+    std::string resultStr;
+    std::string errorStr;
+
+    // http://tizen.org/system/model_name
+    ret = system_info_get_platform_string("http://tizen.org/system/model_name",
+                                          &value);
+    if (ret != SYSTEM_INFO_ERROR_NONE) {
+      resultStr = get_error_message(ret);
```

Comment:
`hasResult` will be true if any one of the items to get information succeeds. If it fails to obtain information from this item, this error message can be displayed as a normal value. It is difficult for users to distinguish between normal value and error message. In error situations, it seems good to exclude this item or provide an empty string.

Reply:
>  If it fails to obtain information for this item,

Can you tell me what is that case? (error message sample?)

Reply:
@bwikbs for example, if it fails to get modelName, but all other values succeeded, the result would be:
```json
 modelName: "some error message", //  (in case of failure getting info from CAPI)
 cpuArch: "armv7l",
 platformVersion: "6.5.0",
 ....
```

Reply:
Ah.. Now I got point! 
At first, We handled an error, but when we did that, the test failed, so it came to this.
We also considered empty, but then there is no way to show the error message..

Reply:
@bwikbs Well.. IMO, this function is not a critical function to send an error message directly to the user. We just need to return as much information as we can return. Rather, it is more useful to let the user check that the value is returned correctly. In the current impl, it is absolutely necessary to check whether the returned value is an error message or not.

Reply:
How about just logging error message instead of sending it?

Reply:
@WonyoungChoi I also feel that your intellectual part is right. But I don't have a opinion because I'm not sure how to do it right.  :smile:

Reply:
It's always the best practice to follow implementations of other platforms.

Reply:
@WonyoungChoi, @bwikbs
Thanks for the suggestion and the history of the current implementation.

@swift-kim
I quickly skimmed the implementations of other platforms, they seem to return null for fields that may not be available in some systems, but assumes that it's always possible to return data for non-nullable fields. I will provide more precise information after looking into it.

Reply:
Based on the implementations of other platforms---specifically Android and Linux---I think it's best to send an empty string for unavailable values like @WonyoungChoi suggessted, and log error messages to Console.

As @bwikbs said, then there is no way to send the error message to dart code, but I think platform interface was designed so that it doesn't care why the values were unavailable, it just understands that some values may be missing and handles them appropriately. For example, by setting default values for non-nullable fields.

Since we implement our own `DeviceInfoPlatform`, we can make dart code receive error messages if we use `BasicMethodChannel` instead of `MethodChannel`, but I don't see the need to make the code different from `device_info_plus` just for this reason.

## Suggestion

As-is

```cpp
ret = system_info_get_platform_string("http://tizen.org/system/model_name",
                                        &value);
if (ret != SYSTEM_INFO_ERROR_NONE) {
    resultStr = get_error_message(ret);
    errorStr.append(" Failed to get model name: ");
    errorStr.append(resultStr);
} else {
    hasResult = true;
    resultStr = std::string(value);
    LOG_INFO("model name : %s\n", value);
    free(value);
}
msg.insert(std::pair<flutter::EncodableValue, flutter::EncodableValue>(
    "modelName", resultStr));
```

To-be

```cpp
ret = system_info_get_platform_string("http://tizen.org/system/model_name",
                                        &value);
if (ret == SYSTEM_INFO_ERROR_NONE) {
    resultStr = std::string(value);
    free(value);
} else {
    resultStr = "";
    LOG_ERROR("Failed to get modelName from the system: %s", get_error_message(ret));
}

msg.insert(std::pair<flutter::EncodableValue, flutter::EncodableValue>(
    "modelName", resultStr));
```

On a different note, since we're already making code changes, how about this to make the code shorter?

```cpp
std::vector<std::pair<std::string, std::string>> requests = {
      {"modelName", "http://tizen.org/system/model_name"},
      {"cpuArch", "http://tizen.org/feature/platform.core.cpu.arch"},
      {"nativeApiVersion",
       "http://tizen.org/feature/platform.native.api.version"},
      {"platformVersion", "http://tizen.org/feature/platform.version"},
      {"webApiVersion", "http://tizen.org/feature/platform.web.api.version"},
      {"buildDate", "http://tizen.org/system/build.date"},
      {"buildId", "http://tizen.org/system/build.id"},
      {"buildString", "http://tizen.org/system/build.string"},
      {"buildTime", "http://tizen.org/system/build.time"},
      {"buildType", "http://tizen.org/system/build.type"},
      {"buildVariant", "http://tizen.org/system/build.variant"},
      {"buildRelease", "http://tizen.org/system/build.release"},
      {"deviceType", "http://tizen.org/system/device_type"},
      {"manufacturer", "http://tizen.org/system/manufacturer"},
  };

for (auto &request : requests) {
    std::string &map_key = request.first;
    std::string &tizen_key = request.second;
    std::string resultStr = "";
    int ret = system_info_get_platform_string(tizen_key.c_str(), &value);
    if (ret == SYSTEM_INFO_ERROR_NONE) {
      resultStr = std::string(value);
      free(value);
    } else {
      LOG_ERROR("Failed to get %s from the system: %s", tizen_key.c_str(),
              get_error_message(ret));
    }
    msg.insert(std::pair<flutter::EncodableValue, flutter::EncodableValue>(
        map_key, resultStr));
}

result->Success(flutter::EncodableValue(msg));
```

Reply:
Wow It's really impressive! :scream: 
Thanks for the detailed explanation! I love it :heart:
It seems to be much better than before.